File: app/src/main/java/cloud/app/vvf/network/api/trakt/AppTrakt.java
package cloud.app.vvf.network.api.trakt;

import android.content.Context;

import androidx.annotation.Nullable;

import com.uwetrottmann.trakt5.TraktV2;
import com.uwetrottmann.trakt5.entities.TraktError;
import com.uwetrottmann.trakt5.entities.TraktOAuthError;

import cloud.app.vvf.common.settings.PrefSettings;
import cloud.app.vvf.network.api.trakt.services.ExtendService;
import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Response;


public class AppTrakt extends TraktV2 {

    public static String CALLBACK_URL = "mmat://trakt/auth/callback";
    private final OkHttpClient okHttpClient;
    private final PrefSettings prefSettings;

    public AppTrakt(OkHttpClient okHttpClient, String client_id, String clientSecret, PrefSettings prefSettings) {
        super(client_id, clientSecret, CALLBACK_URL);
        OkHttpClient.Builder builder = okHttpClient.newBuilder();
        setOkHttpClientDefaults(builder);
        this.okHttpClient = builder.build();
        this.prefSettings = prefSettings;
    }

    /**
     * Check if the request was unauthorized.
     *
     * @see #isUnauthorized(Context, Response)
     */
    public static boolean isUnauthorized(Response<?> response) {
        return response.code() == 401;
    }

    /**
     * Check if the associated Trakt account is locked.
     * <p>
     * https://trakt.docs.apiary.io/#introduction/locked-user-account
     */
    public static boolean isAccountLocked(Response<?> response) {
        return response.code() == 423;
    }

    public static boolean isUnauthorized(Context context, Response<?> response) {
        // current access token is invalid, remove it and notify user to re-connect
        //TraktOAuthSettings.setCredentialsInvalid();
        return response.code() == 401;
    }

    @Nullable
    public static String checkForTraktError(TraktV2 trakt, Response<?> response) {
        TraktError error = trakt.checkForTraktError(response);
        if (error != null && error.message != null) {
            return error.message;
        } else {
            return null;
        }
    }

    @Nullable
    public static String checkForTraktOAuthError(TraktV2 trakt, Response<?> response) {
        TraktOAuthError error = trakt.checkForTraktOAuthError(response);
        if (error != null && error.error != null && error.error_description != null) {
            return error.error + " " + error.error_description;
        } else {
            return null;
        }
    }

    /**
     * Executes the given call. Will return null if the call fails for any reason, including auth
     * failures.
     */
    public static <T> T executeCall(Call<T> call, String action) {
        try {
            Response<T> response = call.execute();
            if (response.isSuccessful()) {
                return response.body();
            } else {
                //Errors.logAndReport(action, response);
            }
        } catch (Exception e) {
            //Errors.logAndReport(action, e);
        }
        return null;
    }

    /**
     * Executes the given call. If the call fails because auth is invalid, removes the current
     * access token and displays a warning notification to the user.
     */

    public static <T> T executeAuthenticatedCall(Context context, Call<T> call, String action) {
        try {
            Response<T> response = call.execute();
            if (response.isSuccessful()) {
                return response.body();
            } else {
                if (!isUnauthorized(context, response)) {
                    //Errors.logAndReport(action, response);
                }
            }
        } catch (Exception e) {
            // Errors.logAndReport(action, e);
        }
        return null;
    }

    @Override
    public String accessToken() {
        return TraktOAuthSettings.getAccessToken(prefSettings);
    }

    @Override
    public String refreshToken() {
        return TraktOAuthSettings.getRefreshToken(prefSettings);
    }

    protected synchronized OkHttpClient okHttpClient() {

        return okHttpClient;
    }

    public ExtendService extendService() {
        return retrofit().create(ExtendService.class);
    }
}

File: app/src/main/java/cloud/app/vvf/extension/tmdb/services/trakt/AppTrakt.kt
package cloud.app.vvf.extension.tmdb.services.trakt

import android.content.Context
import cloud.app.vvf.common.settings.PrefSettings
import cloud.app.vvf.extension.tmdb.services.trakt.services.ExtendService
import com.uwetrottmann.trakt5.TraktV2
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Response

class AppTrakt(
    okHttpClient: OkHttpClient,
    client_id: String,
    clientSecret: String,
    prefSettings: PrefSettings
) : TraktV2(client_id, clientSecret, CALLBACK_URL) {
    private val okHttpClient: OkHttpClient
    private val prefSettings: PrefSettings

    init {
        val builder = okHttpClient.newBuilder()
        setOkHttpClientDefaults(builder)
        this.okHttpClient = builder.build()
        this.prefSettings = prefSettings
    }

    override fun accessToken(): String? {
        return TraktOAuthSettings.getAccessToken(prefSettings)
    }

    override fun refreshToken(): String? {
        return TraktOAuthSettings.getRefreshToken(prefSettings)
    }

    @Synchronized
    override fun okHttpClient(): OkHttpClient {
        return okHttpClient
    }

    fun extendService(): ExtendService {
        return retrofit().create<ExtendService>(ExtendService::class.java)
    }

    companion object {
        var CALLBACK_URL: String = "mmat://trakt/auth/callback"

        /**
         * Check if the request was unauthorized.
         *
         * @see .isUnauthorized
         */
        fun isUnauthorized(response: Response<*>): Boolean {
            return response.code() == 401
        }

        /**
         * Check if the associated Trakt account is locked.
         *
         *
         * https://trakt.docs.apiary.io/#introduction/locked-user-account
         */
        fun isAccountLocked(response: Response<*>): Boolean {
            return response.code() == 423
        }

        fun isUnauthorized(context: Context?, response: Response<*>): Boolean {
            // current access token is invalid, remove it and notify user to re-connect
            //TraktOAuthSettings.setCredentialsInvalid();
            return response.code() == 401
        }

        fun checkForTraktError(trakt: TraktV2, response: Response<*>): String? {
            val error = trakt.checkForTraktError(response)
            if (error != null && error.message != null) {
                return error.message
            } else {
                return null
            }
        }

        fun checkForTraktOAuthError(trakt: TraktV2, response: Response<*>): String? {
            val error = trakt.checkForTraktOAuthError(response)
            if (error != null && error.error != null && error.error_description != null) {
                return error.error + " " + error.error_description
            } else {
                return null
            }
        }

        /**
         * Executes the given call. Will return null if the call fails for any reason, including auth
         * failures.
         */
        fun <T> executeCall(call: Call<T?>, action: String?): T? {
            try {
                val response = call.execute()
                if (response.isSuccessful()) {
                    return response.body()
                } else {
                    //Errors.logAndReport(action, response);
                }
            } catch (e: Exception) {
                //Errors.logAndReport(action, e);
            }
            return null
        }

        /**
         * Executes the given call. If the call fails because auth is invalid, removes the current
         * access token and displays a warning notification to the user.
         */
        fun <T> executeAuthenticatedCall(context: Context?, call: Call<T?>, action: String?): T? {
            try {
                val response = call.execute()
                if (response.isSuccessful()) {
                    return response.body()
                } else {
                    if (!isUnauthorized(context, response)) {
                        //Errors.logAndReport(action, response);
                    }
                }
            } catch (e: Exception) {
                // Errors.logAndReport(action, e);
            }
            return null
        }
    }
}
